optimum/onnxruntime/modeling_ort.py

```diff
@@ -835,36 +810,37 @@ def _prepare_io_binding(
         """
         io_binding = model.io_binding()
 
-        name_to_np_type = TypeHelper.get_io_numpy_type_map(model)
+        input_shapes = {}
+        for input_name in self.input_names.keys():
+            input_shapes[input_name] = model_inputs[input_name].shape
 
-        input_name_to_shape = {}
-        for idx, tensor in enumerate(model_inputs):
-            if tensor is None:
-                continue
-            name = ordered_input_names[idx]
-            tensor = tensor.contiguous()
-            input_name_to_shape[name] = tensor.shape
+            if not model_inputs[input_name].is_contiguous():
+                model_inputs[input_name] = model_inputs[input_name].contiguous()
+
+            tensor_dtype = model_inputs[input_name].dtype
+            expected_dtype = TypeHelper.ort_type_to_torch_type(self.input_dtypes[input_name])
+            if tensor_dtype != expected_dtype:
+                model_inputs[input_name] = model_inputs[input_name].to(expected_dtype)
 
-            data_ptr = tensor.data_ptr()
-            if "past" in name and data_ptr == 0:
-                # During first generation, sequence_length can be 0 when use_cache=True, which results in data_ptr to also be 0.
-                # To keep compatibility with IO binding, we pass the data pointer of input_ids instead. This will have no impact because past_key_values will not be used during the first generation.
-                data_ptr = model_inputs[0].data_ptr()
+            data_ptr = model_inputs[input_name].data_ptr()
+            if data_ptr == 0:
+                data_ptr = model_inputs["input_ids"].data_ptr()
```

Comment:
```suggestion
            if data_ptr == 0:
                # During first generation, sequence_length can be 0 when use_cache=True, which results in data_ptr to also be 0.
                # To keep compatibility with IO binding, we pass the data pointer of input_ids instead. This will have no impact because past_key_values will not be used during the first generation.
                data_ptr = model_inputs["input_ids"].data_ptr()
```